navit/navit.c

```diff
@@ -3504,10 +3505,29 @@ void navit_layout_switch(struct navit *n) {
                 && valid_attr.u.num==attr_position_valid_invalid) {
             return; //No valid fix yet
         }
+
         if (vehicle_get_attr(n->vehicle->vehicle, attr_position_coord_geo,&geo_attr,NULL)!=1) {
             //No position - no sun
             return;
         }
+
+        if (streetflags != NULL) {
+
+            dbg(lvl_debug, "streetflags: %i", *streetflags);
+
+            if (*streetflags & AF_UNDERGROUND) {
+                // We are in a tunnel and if we have a nightlayout -> switch to nightlayout
+                if (l->nightname) {
+                    navit_set_layout_by_name(n, l->nightname);
+                    dbg(lvl_debug, "tunnel nightlayout");
+                    return;
```

Comment:
OK, ran it and found another problem: navit_layout_switch contains code that should result in navit not calculating the layout switch more often than once per minute. See check at line 3497. Since your addition does not correctly update n->prevTs, this leads to the funny situation that the switch when entering a tunnel is delayed, the switch when leaving the tunnel usually is not as n->prevTs is not updated and stays older then 60s.
IMHO the function should be refactored as a whole, but I don't want to request doing so here. But to get this in, please treat n->prevTs right.

For me I don't care if tunnel switching should not be delayed, or will be delayed as well, but at least it should behave symmetrically when entering or leaving a tunnel.

Reply:
And remove the "todo" mark, as this change aims to resolve it.

Reply:
Seems OK now.